pkg/virt-controller/watch/vmi.go

```diff
@@ -1039,3 +1107,355 @@ func (c *VMIController) currentPod(vmi *virtv1.VirtualMachineInstance) (*k8sv1.P
+	oldStatusMap := make(map[string]virtv1.VolumeStatus)
+	for _, status := range oldStatus {
+		oldStatusMap[status.Name] = status
+	}
+
+	hotplugVolumes := c.getHotplugVolumes(vmi, virtlauncherPod)
+	hotplugVolumesMap := make(map[string]*virtv1.Volume)
+	for _, volume := range hotplugVolumes {
+		hotplugVolumesMap[volume.Name] = volume
+	}
+
+	attachmentPods, err := c.virtlauncherAttachmentPods(virtlauncherPod)
+	if err != nil {
+		return err
+	}
+	newStatus := make([]virtv1.VolumeStatus, 0)
+	for _, volume := range vmi.Spec.Volumes {
+		status := virtv1.VolumeStatus{}
+		if _, ok := oldStatusMap[volume.Name]; ok {
+			// Already have the status, modify if needed
+			status = oldStatusMap[volume.Name]
+		} else {
+			status.Name = volume.Name
+		}
+		// Remove from map so I can detect existing volumes that have been removed from spec.
+		delete(oldStatusMap, volume.Name)
+		if _, ok := hotplugVolumesMap[volume.Name]; ok {
+			// Hotplugged volume
+			if status.HotplugVolume == nil {
+				status.HotplugVolume = &virtv1.HotplugVolumeStatus{}
+			}
+			attachmentPod := c.findAttachmentPodByVolumeName(volume.Name, attachmentPods)
+			if attachmentPod == nil {
+				status.HotplugVolume.AttachPodName = ""
+				status.HotplugVolume.AttachPodUID = ""
+				// Pod is gone, or hasn't been created yet, check for the PVC associated with the volume to set phase and message
+				phase, reason, message := c.getVolumePhaseMessageReason(&volume, vmi.Namespace)
+				status.Phase = phase
+				status.Message = message
+				status.Reason = reason
+			} else {
+				status.HotplugVolume.AttachPodName = attachmentPod.Name
+				status.HotplugVolume.AttachPodUID = attachmentPod.UID
+				if c.canMoveToAttachedPhase(status.Phase) {
+					status.Phase = virtv1.HotplugVolumeAttachedToNode
+					status.Message = fmt.Sprintf("Created hotplug attachment pod %s, for volume %s", attachmentPod.Name, volume.Name)
+					status.Reason = SuccessfulCreatePodReason
+				}
+			}
+		}
+		newStatus = append(newStatus, status)
+	}
+
+	// We have updated the status of current volumes, but if a volume was removed, we want to keep that status, until there is no
+	// associated pod, then remove it. Any statuses left in the map are statuses without a matching volume in the spec.
+	for k, v := range oldStatusMap {
+		attachmentPod := c.findAttachmentPodByVolumeName(k, attachmentPods)
+		if attachmentPod != nil {
+			v.HotplugVolume.AttachPodName = attachmentPod.Name
+			v.HotplugVolume.AttachPodUID = attachmentPod.UID
+			v.Phase = virtv1.HotplugVolumeDetaching
+			if attachmentPod.DeletionTimestamp != nil {
+				v.Message = fmt.Sprintf("Deleted hotplug attachment pod %s, for volume %s", attachmentPod.Name, k)
+				v.Reason = SuccessfulDeletePodReason
+			}
+			// If the pod exists, we keep the status.
+			newStatus = append(newStatus, v)
+		}
+	}
+	sort.SliceStable(newStatus, func(i, j int) bool {
+		return strings.Compare(newStatus[i].Name, newStatus[j].Name) == -1
+	})
```

Comment:
interesting, so this sort guarantees a predictable ordering. I like that. If virt-controller and virt-handler observe the same ordering, then we won't have to worry about either of them attempting to compete by shuffling the order somehow.

Reply:
Exactly since the volume names are unique, I can guarantee that.